src/app/(app)/product-list.tsx

```diff
@@ -1,95 +1,191 @@
+          data={products}
+          onScroll={onScroll}
+          renderItem={({ item, index }) => {
+            const isFirstItem = index === 0;
+            const isLastItem = index === products.length - 1;
+
+            return (
+              <View
+                className={`mx-4 border ${isFirstItem ? 'rounded-t-lg' : ''} ${
+                  isLastItem ? 'rounded-b-lg' : ''
+                }`}
+              >
+                <ProductCard
+                  id={item.id}
+                  price={item.unit_price}
+                  state={item.state}
+                  name={item.title}
+                  image_url={item.pictures[0]}
+                />
+              </View>
+            );
+          }}
+          keyExtractor={(item) => item.id.toString()}
+        />
+      </View>
+    );
+  }
+);
+
+const useProductPagination = (
+  page: number,
+  data: FetchProductsResponse | undefined,
+  setPage: React.Dispatch<React.SetStateAction<number>>
+) => {
+  const handleEndReached = useCallback(() => {
+    if (page * 7 - 1 < (data?.pagination?.count ?? 0)) {
+      setPage((prev) => prev + 1);
+    }
+  }, [page, data, setPage]);
+
+  const handleScroll = useCallback(
+    ({ nativeEvent }: any) => {
+      const { layoutMeasurement, contentOffset, contentSize } = nativeEvent;
+      if (
+        layoutMeasurement.height + contentOffset.y >=
+        contentSize.height - 20
+      ) {
+        handleEndReached();
+      }
+    },
+    [handleEndReached]
+  );
+
+  return { handleEndReached, handleScroll };
+};
+
+export default function ProducList() {
+  const [page, setPage] = useState(1);
+  const [products, setProducts] = useState<Product[]>([]);
+  const { data } = useProducts({ variables: { page: page, items: 7 } });
+  const [fetchedProducts, setFetchedProducts] = useState<Product[]>([]);
+  const [query, setQuery] = useState<string>('');
+  const [shouldReset, setShouldReset] = useState(false);
+  const { handleEndReached, handleScroll } = useProductPagination(
+    page,
+    data,
+    setPage
+  );
+
+  const resetSearchBar = () => {
+    setProducts(fetchedProducts);
+    setShouldReset(true);
+    setQuery('');
+    setTimeout(() => setShouldReset(false), 100);
+  };
+
+  const handleProductSelect = (searchQuery: string) => {
+    if (searchQuery === '') {
+      resetSearchBar();
+      return;
+    }
+    const queryProducts = fetchedProducts.filter((product) =>
+      product.title.toLowerCase().includes(searchQuery.toLowerCase())
+    );
+    setQuery(searchQuery);
+    setProducts(queryProducts.slice(0, 7));
+  };
+
+  useEffect(() => {
+    if (data?.data) {
+      setFetchedProducts((prev) => [...prev, ...data.data]);
+    }
+  }, [data]);
+
+  useEffect(() => setProducts(fetchedProducts), [fetchedProducts]);
+
+  return (
+    <SafeAreaView className="flex-1">
+      <HeaderLogo />
+      <SearchBar
+        onProductSelect={handleProductSelect}
+        cleanQuery={shouldReset}
+      />
```

Comment:
You're already maintaining the value of the search term in the `query` state value at this screen, so you could pass it down to SearchBar and update the searchTerm inside the component instead of having to handle an extra cleanQuery state and the timeout effect to reset it.

OnProductSelect is not an intuitive name for what the function does, and same with the property name.

What about something like onSubmit={filterProducts} ?

This will improve readability and maintainability, there's no better documentation that the code itself :D

Reply:
I want to trigger the function of search just on the submit, if i pass the value it triggers the function of search on each change in query. Is there some way to avoid this behavior?